### PR TITLE
monitoring: migrate from OpenCensus to OpenTelemetry

### LINKED ISSUE #44513: telemetry: move to OpenTelemetry lib
[opencensus-go](https://github.com/census-instrumentation/opencensus-go) announced for deprecated for a long time, should Istio move to use OpenTelemetry lib?

see project status [here](https://github.com/open-telemetry/opentelemetry-go#project-status)

cc @howardjohn @lei-tang @kyessenov

### PR DESCRIPTION
**Pull request by howardjohn**
~Blocked by https://github.com/open-telemetry/opentelemetry-go/pull/4306. I have a `hold` until that is ready; should be fairly quick~

For https://github.com/istio/istio/issues/44743

Fixes https://github.com/istio/istio/issues/44513

This PR migrates from OpenCensus to OpenTelemetry for Go metrics. This
doesn't impact proxy metrics, or non-metrics observability.

These changes are intended to be 100% transparent to end users.

The changes are *mostly* transparent from `pkg/monitoring` public
interface, with a few changes:
* `Register` is no longer needed. Creating the metric registers it. This
  makes no real difference in any of our usage
* MustCreateLabel -> CreateLabel; it is infallable
* No need to define the set of keys upfront with the metric.
* RegisterIf -> WithEnabled
* monitortest behaves slightly different; only metrics recorded after
  its creation are counted. A few tests are updated to handle this
change.
* Only one exporter can be defined; this only impacts tests, which are
  updated to handle this.

There are a number of changes, of course, to the internals. Most of this
is a fairly mechanical translation. A few quirks:

* There is no real Gauge like in OC. Instead, we can only do a callback.
  So we need to handle a bit of this on our end; not a huge deal, and
looks like it will be added eventually.
* Setting histogram buckets is a bit awkward currently. This will be
  addressed eventually upstream.
* monitortest previously was based on OC internals. This has been re-written to *prometheus* instead of OTEL. This is because prometheus is our primary usage, so its a bit more realistic.

~This results in a 1% increase in the binary size, which is due to importing both OC and Otel. This is addressed in https://github.com/istio/istio/pull/46010, which introduces a 1% decrease in binary size _when combined with this PR_.~

Recording metrics is now *much* cheaper. This may allow us to start enabling metrics that were previously disabled (TBD, I am planning a more thorough design of our overall observability, would prefer to avoid changes until then):

```
name                      old time/op    new time/op    delta
Counter/no_labels-7          295ns ± 3%      56ns ±10%   -80.91%  (p=0.008 n=5+5)
Counter/dynamic_labels-7     870ns ±26%     434ns ± 4%   -50.16%  (p=0.008 n=5+5)
Counter/static_labels-7      407ns ± 2%     122ns ± 1%   -69.98%  (p=0.029 n=4+4)

name                      old alloc/op   new alloc/op   delta
Counter/no_labels-7          68.0B ± 0%      0.0B       -100.00%  (p=0.008 n=5+5)
Counter/dynamic_labels-7      760B ± 0%      320B ± 0%      ~     (p=0.079 n=4+5)
Counter/static_labels-7      80.0B ± 0%      0.0B       -100.00%  (p=0.008 n=5+5)

name                      old allocs/op  new allocs/op  delta
Counter/no_labels-7           3.00 ± 0%      0.00       -100.00%  (p=0.008 n=5+5)
Counter/dynamic_labels-7      13.0 ± 0%       6.0 ± 0%   -53.85%  (p=0.008 n=5+5)
Counter/static_labels-7       3.00 ± 0%      0.00       -100.00%  (p=0.008 n=5+5)
```

### CODE REVIEW

**Comment by zirain** [comment]
FYI: I test with opentelemetry-go offline, it's will add `_total` suffix fro counter, which maybe a regression

https://github.com/open-telemetry/opentelemetry-go/blob/4d473d105a5f59a58d153e7c492a4ea0ee027871/exporters/prometheus/exporter.go#L216

File: pkg/monitoring/monitoring.go
Comment (zirain) [inline pkg/monitoring/monitoring.go:334]: this looks odd, maybe should review the design of Metric interface?

Comment (howardjohn) [reply pkg/monitoring/monitoring.go:334]: The alternative is to have 3 concrete implementations. Which we could do but I don't think this is so bad?

I guess we could have 4 structs - a  common one with common logic and 3 for (counter,guage,distribution)? I can see if that is cleaner.

BTW this is all somewhat hacky, I was planning to clean it up if we move forward

Comment (zirain) [reply pkg/monitoring/monitoring.go:334]: > The alternative is to have 3 concrete implementations.

I try this way, it makes PR too huge.

File: common/config/.golangci.yml
Comment (zirain) [inline common/config/.golangci.yml:287]: you should not use opencensus directly at the same time

File: security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient_test.go
Comment (zirain) [inline security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient_test.go:79]: is this worth a release notes?

**Review by zirain — APPROVED** [review]
Great work! Thanks @howardjohn

**Review by ramaraochavali — APPROVED** [review]

File: security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient_test.go
Comment (howardjohn) [reply security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient_test.go:79]: Stale comment  - its not broken anymore. Let me clean this up

**Review by ericvn — APPROVED** [review]

**Comment by costinm** [comment]
Looks very good.

One concern - not blocking this PR - is that I would prefer if labels are declared at metric creation, even if otel doesn't require it.

File: cni/pkg/install/monitoring.go
Comment (costinm) [inline cni/pkg/install/monitoring.go:29]: I would not remove this part ( maybe refactor it to avoid the MustCreateLabel - but having a list of labels is very informative, and will still be required if we want monitoring to have a prometheus implementation. 

If we want OTel only ( which I don't mind ) - we don't need an abstraction/wrapper on top of the SDK. But even then it would be good to keep track of what labels are generated, it is probably the most critical aspect of a metric.

File: operator/pkg/metrics/monitoring.go
Comment (costinm) [inline operator/pkg/metrics/monitoring.go:46]: Where are they used - I see WithLabels is removed.

File: operator/pkg/metrics/monitoring.go
Comment (costinm) [inline operator/pkg/metrics/monitoring.go:205]: What does it do - isn't it another must register ?

File: pilot/cmd/pilot-agent/status/server.go
Comment (costinm) [inline pilot/cmd/pilot-agent/status/server.go:152]: If we're doing this - shouldn't agent also use otel sdk ? Not in this PR, of course - but a TODO and comment explaining why prometheus is used directly would help other developers.

File: pilot/pkg/config/kube/crdclient/metrics.go
Comment (costinm) [inline pilot/pkg/config/kube/crdclient/metrics.go:22]: Not entirely sure why this was renamed - either remove or leave it as is, we can live with 'Must' for a while...

**Review by costinm — APPROVED** [review]
I am approving the PR - not entirely happy, but it is clearly a huge improvement (opencensus is long deprecated and unsupported).

I think we need to follow up - and remove the wrappers, if we use otel let's use the otel API without an istio abstraction that just confuses people. And add back some info about the labels for each metris - in the description or somewhere, we also probably need to generate docs for each metric. 

All of this can be done in separate PRs and incrementally.

I would also add some doc - and make it clear if indeed the intent is to start using otel API and if it is ok for new metrics ( and gradually for the old ones ) to replace the use of the wrapper with the proper API.

**Pull request merged**
